### PR TITLE
Add info about MSBuild /pp option to show the whole project as seen by MSBuild

### PR DESCRIPTION
**Pull request by bleroy**
With the new simplified csproj format, a lot of information that used to be explicit has been pushed into the SDK and its targets. As a result, it's hard to track some of them. This change shows how to see the whole csproj as MSBuild sees it.

## Suggested Reviewers

@mairaw @ellismg

### CODE REVIEW

File: docs/core/tools/csproj.md [hunk 1]
```diff
@@ -69,6 +69,13 @@ This change does not modify the main mechanics of other includes. However, if yo
 ### Recommendation
 With csproj, we recommend that you remove the default globs from your project and only add file paths with globs for those artifacts that your app/library needs for various scenarios (runtime, NuGet packaging, etc.)
 
+## How to see the whole project as MSBuild sees it
```
Comment (mairaw) [inline docs/core/tools/csproj.md]: update ms.date on these two topics to reflect the topic has been updated

File: docs/core/tools/csproj.md [hunk 2]
```diff
@@ -69,6 +69,13 @@ This change does not modify the main mechanics of other includes. However, if yo
 ### Recommendation
 With csproj, we recommend that you remove the default globs from your project and only add file paths with globs for those artifacts that your app/library needs for various scenarios (runtime, NuGet packaging, etc.)
```
Comment (GuardRex) [inline docs/core/tools/csproj.md]: Not part of the PR but just noting here in passing: "etc." is forbidden. It should go along the lines of ...

> various scenarios (for example, runtime and NuGet packaging).

(The current line is missing a period at the end, too.)

File: docs/core/tools/csproj.md [hunk 3]
```diff
@@ -69,6 +69,13 @@ This change does not modify the main mechanics of other includes. However, if yo
...
 
+## How to see the whole project as MSBuild sees it
+
+While those csproj changes greatly simplify project files, it is sometimes useful to be able to look at the fully expanded project as MSBuild sees it once the SDK and its targets are included. This can be done by simply typing the following in the command-line:
```
Comment (mairaw) [inline docs/core/tools/csproj.md]: it is sometimes useful to be able to look at the -> can you simplify this sentence? perhaps "you might wanna see the"

This can be done by simply typing the following in the command-line -> This can be done by specifying the `/pp` switch to the [dotnet msbuild](dotnet-msbuild.md) command, such as in the following example:

My suggestion is based on the fact that you can also pass a filepath to the /pp switch, no? (although I'm having some issues using dotnet msbuild /pp:[filepath]. It's doing some weird stuff. I might file a CLI bug on this once I confirm the behavior.

File: docs/core/tools/csproj.md [hunk 4]
```diff
@@ -69,6 +69,13 @@ This change does not modify the main mechanics of other includes. However, if yo
...
+
+```
+dotnet msbuild /pp
+```
```
Comment (svick) [inline docs/core/tools/csproj.md]: When I tried this on one of my projects, it didn't work (the output was just: `coreapp -> C:\code\tmp\coreapp\bin\Debug\netcoreapp1.1\coreapp.dll`). It took me a while to figure out it was because I had a .sln file in the directory.

Is that something that's worth mentioning here?

Also, is that the expected behavior?

File: docs/core/tools/csproj.md (hunk 3 above)
Comment (GuardRex) [inline docs/core/tools/csproj.md]: "it is" :point_right: "it's"
There's a general leaning away from "you can" language. Also, "command-line" :point_right: "command prompt." Would you consider expanding a bit on ...

> This can be done by simply typing the following in the command-line:

... with something along the lines of this, which explains out a bit more what's going to happen? ....

> Preprocess the project with the following command and switch, which shows which files are imported, their sources, and their contributions to the build without actually building the project:

(:point_up: Shamelessly stolen sentiment from the MSDN topic.)

**Review by guardrex — CHANGES_REQUESTED** [review]
Hi @belroy ... a few minor suggestions.

File: docs/core/tools/csproj.md (hunk 4 above)
Comment (bleroy) [inline docs/core/tools/csproj.md]: mmm, no, it doesn't sound right. I'll ask.

File: docs/core/tools/csproj.md (hunk 2 above)
Comment (bleroy) [inline docs/core/tools/csproj.md]: That is not part of this pull request. I'm fine fixing it here, but it doesn't sound right.

**Review by mairaw — COMMENTED** [review]
a few comments

File: docs/core/tools/csproj.md (hunk 2 above)
Comment (GuardRex) [inline docs/core/tools/csproj.md]: We've had an on-going, occasional background discussion about fixing other little probs as we go. Yes, it's not part of this ... it just looks bad ... these little grammatical problems. Sure, I'm cool either way. We can nab this on another PR if you like.

Comment (mairaw) [inline docs/core/tools/csproj.md]: I just learned a new style guide rule. 😄

Comment (GuardRex) [inline docs/core/tools/csproj.md]: I feel like I'm doing that hourly!!! lol

**Review by guardrex — APPROVED** [review]

File: docs/core/tools/csproj.md (hunk 4 above)
Comment (bleroy) [inline docs/core/tools/csproj.md]: Yup, this is a bug. Filed here: https://github.com/Microsoft/msbuild/issues/2131

Thanks for the heads up.

File: docs/core/tools/csproj.md [hunk 5]
```diff
@@ -67,8 +67,17 @@ Setting this property to `false` will override implicit inclusion and the behavi
...
 
 ### Recommendation
-With csproj, we recommend that you remove the default globs from your project and only add file paths with globs for those artifacts that your app/library needs for various scenarios (runtime, NuGet packaging, etc.)
+With csproj, we recommend that you remove the default globs from your project and only add file paths with globs for those artifacts that your app/library needs for various scenarios (for example, runtime and NuGet packaging).
 
+## How to see the whole project as MSBuild sees it
+
+While those csproj changes greatly simplify project files, you might want to see the fully expanded project as MSBuild sees it once the SDK and its targets are included. Preprocess the project with the `/pp` switch of the [`dotnet msbuild`](dotnet-msbuild.md) command, which shows which files are imported, their sources, and their contributions to the build without actually building the project:
+
+`dotnet msbuild /pp`
```
Comment (rainersigwald) [inline docs/core/tools/csproj.md]: Honestly, this form is so chatty that I wouldn't recommend even mentioning it--just always specify a path.

File: docs/core/tools/csproj.md [hunk 6]
```diff
@@ -67,8 +67,17 @@ Setting this property to `false` will override implicit inclusion and the behavi
 This change does not modify the main mechanics of other includes. However, if you wish to specify, for example, some files to get published with your app, you can still use the known mechanisms in *csproj* for that (for example, the `<Content>` element).
 
 ### Recommendation
-With csproj, we recommend that you remove the default globs from your project and only add file paths with globs for those artifacts that your app/library needs for various scenarios (runtime, NuGet packaging, etc.)
+With csproj, we recommend that you remove the default globs from your project and only add file paths with globs for those artifacts that your app/library needs for various scenarios (for example, runtime and NuGet packaging).
 
+## How to see the whole project as MSBuild sees it
+
+While those csproj changes greatly simplify project files, you might want to see the fully expanded project as MSBuild sees it once the SDK and its targets are included. Preprocess the project with the `/pp` switch of the [`dotnet msbuild`](dotnet-msbuild.md) command, which shows which files are imported, their sources, and their contributions to the build without actually building the project:
```
Comment (rainersigwald) [inline docs/core/tools/csproj.md]: Too bad it's hard to link to the `/pp` section on https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-command-line-reference.

File: docs/core/tools/csproj.md [hunk 7]
```diff
@@ -67,8 +67,17 @@ Setting this property to `false` will override implicit inclusion and the behavi
 This change does not modify the main mechanics of other includes. However, if you wish to specify, for example, some files to get published with your app, you can still use the known mechanisms in *csproj* for that (for example, the `<Content>` element).
 
 ### Recommendation
-With csproj, we recommend that you remove the default globs from your project and only add file paths with globs for those artifacts that your app/library needs for various scenarios (runtime, NuGet packaging, etc.)
+With csproj, we recommend that you remove the default globs from your project and only add file paths with globs for those artifacts that your app/library needs for various scenarios (for example, runtime and NuGet packaging).
 
+## How to see the whole project as MSBuild sees it
```
Comment (rainersigwald) [inline docs/core/tools/csproj.md]: One major wrinkle that may be worth mentioning here (it keeps biting me and I should know better than anyone):

If your project has multiple `TargetFramework`**`s`**, just running `dotnet msbuild /pp` on it will preprocess the "outer" build, which isn't very interesting--it just dispatches to the inner builds. You usually want to also specify a TFM, for example with `dotnet msbuild /p:TargetFramework=netcoreapp2.0 /pp`. Note that the convenience `dotnet -f {tfm}` doesn't work, you have to specify the framework as an MSBuild property.

**Review by rainersigwald — COMMENTED** [review]
Some nits from the MSBuild perspective.

File: docs/core/tools/csproj.md (hunk 6 above)
Comment (svick) [inline docs/core/tools/csproj.md]: I think adding an `a name` tag should help with that: https://github.com/svick/visualstudio-docs/blob/patch-1/docs/msbuild/msbuild-command-line-reference.md#preprocess. Should I submit that change as a PR?

File: docs/core/tools/dotnet-msbuild.md [hunk 8]
```diff
@@ -40,4 +40,8 @@ Build a project and its dependencies using Release configuration:
 
 Run the publish target and publish for the `osx.10.11-x64` RID:
 
-`dotnet msbuild /t:Publish /p:RuntimeIdentifiers=osx.10.11-x64`
\ No newline at end of file
+`dotnet msbuild /t:Publish /p:RuntimeIdentifiers=osx.10.11-x64`
+
+See the whole project with all targets included by the SDK:
+
+`dotnet msbuild /pp`
```
Comment (ellismg) [inline docs/core/tools/dotnet-msbuild.md]: Similar to @rainersigwald feed back above, we should probably just include an output filename here.

File: docs/core/tools/csproj.md (hunk 6 above)
Comment (mairaw) [inline docs/core/tools/csproj.md]: That would work @svick. @rainersigwald is /pp:[filepath] supposed to work?

Comment (rainersigwald) [inline docs/core/tools/csproj.md]: @svick Oh, that's awesome!

@mairaw Yes, in my head `/pp:path` is the only way this works. I think I didn't know it would go to stdout with no path specified for several years after I learned about it :)

Comment (svick) [inline docs/core/tools/csproj.md]: Ok, submitted as https://github.com/Microsoft/visualstudio-docs/pull/133.

Comment (mairaw) [inline docs/core/tools/csproj.md]: Ok, that makes sense now. That's the name of the *output* file. It'd help if I read the docs, no? 😊

Comment (rainersigwald) [inline docs/core/tools/csproj.md]: That actually brings up a point maybe worth mentioning here: you can combine `/pp:out` with a `path/to/project` to preprocess _that project_. Or just do it in a directory to get the project that MSBuild would build in that directory.

So both of these are valid:
```
msbuild /pp:preprocessed.xml c:\src\Path\to\library.csproj
msbuild /pp:preprocessed.xml
```

Where the latter will preprocess the project in the current working directory, or error if there's more than one.

That's the same way that all MSBuild invocations work so it probably doesn't need to be mentioned here. But while we're talking about it . . .
